src/dune_engine/dep.ml

```diff
@@ -315,44 +314,33 @@ module Set = struct
     file_selector
       (File_selector.create ~dir File_selector.Predicate_with_id.false_)
 
-  module Source_tree_map_reduce =
-    Source_tree.Dir.Make_map_reduce (Memo) (Monoid.Union (M))
-
-  let source_tree dir =
-    let prefix_with, dir = Path.extract_build_context_dir_exn dir in
-    Source_tree.find_dir dir >>= function
-    | None -> Memo.return empty
-    | Some dir ->
-      Source_tree_map_reduce.map_reduce dir ~traverse:Sub_dirs.Status.Set.all
-        ~f:(fun dir ->
-          let files = Source_tree.Dir.files dir in
-          let path =
-            Path.append_source prefix_with (Source_tree.Dir.path dir)
-          in
+  module Source_tree (Union : sig
+    type 'a result
+
+    val union_all :
+      Path.t -> f:(path:Path.t -> files:Filename.Set.t -> t) -> t result
+  end) =
+  struct
+    let files dir =
+      Union.union_all dir ~f:(fun ~path ~files ->
           match String.Set.is_empty files with
-          | true -> Memo.return (singleton (dir_without_files_dep path))
+          | true -> singleton (dir_without_files_dep path)
           | false ->
-            let paths =
-              String.Set.fold files ~init:Path.Set.empty ~f:(fun fn acc ->
-                  Path.Set.add acc (Path.relative path fn))
-            in
-            Memo.return (add_paths empty paths))
-
-  let source_tree_with_file_set dir =
-    let+ t = source_tree dir in
-    let paths =
-      fold t ~init:Path.Set.empty ~f:(fun dep acc ->
-          match dep with
-          | File f -> Path.Set.add acc f
-          | File_selector fs ->
-            assert (
-              File_selector.Predicate_with_id.equal
-                (File_selector.predicate fs)
-                File_selector.Predicate_with_id.false_);
-            acc
-          | _ -> assert false)
-    in
-    (t, paths)
+            Filename.Set.fold files ~init:empty ~f:(fun fn acc ->
+                add acc (File (Path.relative path fn))))
+  end
+
+  let to_files_set_exn t =
```

Comment:
I just realised that calling this function on the result of `Source_tree.files` is kinda crazy: it essentially undoes whatever `Source_tree.files` did and returns the original set of files, sprinkled with `assert false`s for a little spiciness :D 

No immediate action is required, I guess, but we should clearly clean this up in future, somehow.

Reply:
(Maybe we should add my comment above into the code for future reference.)

Reply:
Indeed it's crazy. I would have liked to get rid of it all by inlining it at the callers and then try to extract the sensible functions. But I have no idea what's being used internally at jst so this might have been too disruptive.

Reply:
> (Maybe we should add my comment above into the code for future reference.)

I'm about to send a PR to fix all this weirdness.

Reply:
> But I have no idea what's being used internally at jst so this might have been too disruptive.

Don't worry about, I'll fix our call sites, if any.

Hopefully, one day most of our patch will be public.